File: app/services/property_evaluation/hedonic.rb
# frozen_string_literal: true

require 'matrix'

module PropertyEvaluation
  # Ruby port of audit-engine-v2's `src/audit_engine/hedonic.py` (OLS
  # hedonic regression) — for the Express estimator. Median ₽/m² over
  # comparables ignores object characteristics; this corrects for area,
  # rooms count, and distance to the subject.
  #
  # Model:
  #   log(price_per_sqm) = β0 + β1·rooms + β2·log(area_sqm) + β3·distance_km + ε
  #
  # OLS via stdlib Matrix#solve_LR. N≤300 finishes in <50 ms with no native
  # extensions (no numo/lapack dep). Returns nil if the sample is too small
  # (<8) or the design matrix is degenerate.
  class Hedonic
    MIN_SAMPLE = 8
    CI_Z_95 = 1.96

    Result = Struct.new(
      :predicted_price_per_sqm,
      :ci_lo_95,
      :ci_hi_95,
      :n_used,
      :r_squared,
      :residual_std,
      :feature_names,
      :coefficients,
      keyword_init: true
    ) do
      def to_h
        {
          predicted_price_per_sqm: predicted_price_per_sqm.round(2),
          ci_lo_95: ci_lo_95.round(2),
          ci_hi_95: ci_hi_95.round(2),
          n_used: n_used,
          r_squared: r_squared.round(4),
          residual_std: residual_std.round(4),
          feature_names: feature_names,
          coefficients: coefficients.map { |c| c.round(6) }
        }
      end
    end

    # comparables: Array<Hash> with keys :rooms, :area, :price_per_sqm, :distance_km
    # target: Hash { rooms:, area:, distance_km: }
    def self.call(comparables:, target:)
      new(comparables, target).call
    end

    def initialize(comparables, target)
      @raw = comparables
      @target = target
    end

    def call
      rows, ys = build_design_matrix
      n = rows.length
      return nil if n < MIN_SAMPLE

      x_full = Matrix.rows(rows)
      kept_idx = drop_constant_columns(x_full)
      return nil if kept_idx.size < 2 # intercept + at least one feature

      x = Matrix.columns(kept_idx.map { |i| x_full.column(i).to_a })
      return nil unless full_rank?(x)

      # OLS: β = (X'X)⁻¹ X'y
      xt = x.transpose
      xtx = xt * x
      return nil unless xtx.regular?

      beta_reduced = xtx.inverse * xt * Matrix.column_vector(ys)
      beta_reduced = beta_reduced.column(0).to_a

      # Residuals, R², residual std
      y_pred = x * Matrix.column_vector(beta_reduced)
      y_pred = y_pred.column(0).to_a
      residuals = ys.each_with_index.map { |yi, i| yi - y_pred[i] }
      ss_res = residuals.sum { |r| r * r }
      mean_y = ys.sum.to_f / ys.length
      ss_tot = ys.sum { |yi| (yi - mean_y) * (yi - mean_y) }
      r_squared = ss_tot.positive? ? 1.0 - (ss_res / ss_tot) : 0.0
      dof = n - kept_idx.size
      residual_std = dof.positive? ? Math.sqrt(ss_res / dof) : 0.0

      # Predict for target on the reduced feature set
      x_target_full = [1.0, @target[:rooms].to_f, Math.log(@target[:area].to_f), @target[:distance_km].to_f]
      x_target = kept_idx.map { |i| x_target_full[i] }
      log_pred = beta_reduced.each_with_index.sum { |b, i| b * x_target[i] }
      pred = Math.exp(log_pred)
      half_log = CI_Z_95 * residual_std
      ci_lo = Math.exp(log_pred - half_log)
      ci_hi = Math.exp(log_pred + half_log)

      feature_names_full = %w[intercept rooms log_area distance_km]
      Result.new(
        predicted_price_per_sqm: pred,
        ci_lo_95: ci_lo,
        ci_hi_95: ci_hi,
        n_used: n,
        r_squared: r_squared,
        residual_std: residual_std,
        feature_names: kept_idx.map { |i| feature_names_full[i] },
        coefficients: beta_reduced
      )
    rescue ExceptionForMatrix::ErrNotRegular, ZeroDivisionError, Math::DomainError => e
      Rails.logger.info("[Hedonic] regression failed: #{e.class} #{e.message}")
      nil
    end

    private

    # Returns [rows, ys] where rows = [[1.0, rooms, log(area), distance_km], ...]
    # and ys = [log(price_per_sqm), ...]. Drops comps with missing/zero
    # area, rooms, or price_per_sqm.
    def build_design_matrix
      rows = []
      ys = []
      @raw.each do |c|
        area = c[:area].to_f
        ppsm = c[:price_per_sqm].to_f
        next if area <= 0 || ppsm <= 0
        next if c[:rooms].nil?

        distance_km = (c[:distance_km] || 0).to_f
        rows << [1.0, c[:rooms].to_f, Math.log(area), distance_km]
        ys << Math.log(ppsm)
      end
      [rows, ys]
    end

    # Returns indices of columns to keep. Intercept (column 0) is always
    # kept. Non-intercept columns with zero variance are dropped.
    def drop_constant_columns(matrix)
      kept = [0]
      (1...matrix.column_count).each do |j|
        col = matrix.column(j).to_a
        mean = col.sum.to_f / col.length
        variance = col.sum { |v| (v - mean) * (v - mean) } / col.length
        kept << j if variance > 1e-9
      end
      kept
    end

    def full_rank?(matrix)
      matrix.rank == matrix.column_count
    rescue StandardError
      false
    end
  end
end

File: services/audit-engine/engine/src/audit_engine/hedonic.py
"""Hedonic regression для оценки справедливой цены ₽/м².

Зачем:
    Простой median ₽/м² по компам игнорирует характеристики объекта
    (площадь, число комнат, расстояние). 60-метровая двушка на 1-м
    этаже не сопоставима с 90-метровой трёшкой на 10-м напрямую.

Модель:
    log(price_per_sqm) = β0 + β1·rooms + β2·log(area_sqm) + β3·distance_km + ε

    OLS через numpy.lstsq. Прогноз для аудитного объекта возвращается
    с 95% доверительным интервалом (на основе residual std).

Когда возвращается None:
    - n_complete < MIN_HEDONIC_SAMPLE (по умолчанию 8) — слишком мало
      объектов с заполненными площадью/комнатами для разумной регрессии.
    - матрица признаков вырождена (все объекты — однокомнатные).

Интеграция:
    `competitor_analyzer.analyze()` опционально вызывает `fit_and_predict`
    и кладёт результат в `CompetitorAnalysis.hedonic`. Median-based анализ
    остаётся как основной — hedonic это дополнительный сигнал.
"""
from __future__ import annotations

import logging
import math
from dataclasses import dataclass
from typing import Iterable, Optional

import numpy as np

from audit_engine.models import Competitor

logger = logging.getLogger(__name__)

MIN_HEDONIC_SAMPLE = 8
CI_Z_95 = 1.96


@dataclass
class HedonicTarget:
    """Параметры аудитного объекта, которые подставляются в модель."""
    rooms: int
    area_sqm: float
    distance_km: float = 0.0  # сам аудитный объект — точка отсчёта


@dataclass
class HedonicResult:
    predicted_price_per_sqm: float
    ci_lo_95: float
    ci_hi_95: float
    n_used: int
    r_squared: float
    residual_std: float
    feature_names: tuple[str, ...]
    coefficients: tuple[float, ...]

    def to_dict(self) -> dict:
        return {
            "predicted_price_per_sqm": round(self.predicted_price_per_sqm, 2),
            "ci_lo_95": round(self.ci_lo_95, 2),
            "ci_hi_95": round(self.ci_hi_95, 2),
            "n_used": self.n_used,
            "r_squared": round(self.r_squared, 4),
            "residual_std": round(self.residual_std, 4),
            "feature_names": list(self.feature_names),
            "coefficients": [round(c, 6) for c in self.coefficients],
        }


def apartment_type_to_rooms(apt_type: Optional[str]) -> Optional[int]:
    """Маппинг 'Studio'/'1BR'/'2BR'/... → число комнат для hedonic-регрессии.

    Возвращает None, если тип не распознан — тогда вызывающий код просто
    пропускает hedonic.
    """
    if not apt_type:
        return None
    s = apt_type.strip().lower()
    if s in ("studio", "студия"):
        return 0
    for prefix in ("1", "2", "3", "4", "5"):
        if s.startswith(prefix):
            try:
                return int(prefix)
            except ValueError:
                pass
    return None


def _drop_constant_columns(X: np.ndarray) -> tuple[np.ndarray, list[int]]:
    """Удалить колонки без вариативности (кроме intercept в позиции 0).

    Возвращает (X_reduced, kept_indices). intercept всегда сохраняется.
    """
    keep = [0]
    for j in range(1, X.shape[1]):
        if X[:, j].std() > 1e-9:
            keep.append(j)
    return X[:, keep], keep


def _build_design_matrix(comps: list[Competitor]) -> tuple[np.ndarray, np.ndarray]:
    """Возвращает X (n × 4) и y (n,). Отбрасывает объекты с пропусками."""
    rows: list[list[float]] = []
    ys: list[float] = []
    for c in comps:
        if c.area_sqm is None or c.rooms is None or c.price_per_sqm <= 0:
            continue
        if c.area_sqm <= 0:
            continue
        distance_km = (c.distance_m or 0) / 1000.0
        rows.append([1.0, float(c.rooms), math.log(c.area_sqm), distance_km])
        ys.append(math.log(c.price_per_sqm))
    return np.array(rows, dtype=float), np.array(ys, dtype=float)


def fit_and_predict(
    target: HedonicTarget,
    competitors: Iterable[Competitor],
) -> Optional[HedonicResult]:
    """Подогнать OLS на компах, предсказать ₽/м² для target.

    Возвращает None, если sample слишком мал или регрессия вырождена.
    """
    comps = list(competitors)
    X_full, y = _build_design_matrix(comps)
    n = X_full.shape[0]
    if n < MIN_HEDONIC_SAMPLE:
        logger.info(
            "hedonic: %d полных объектов < %d — пропускаю регрессию",
            n, MIN_HEDONIC_SAMPLE,
        )
        return None

    # Если какая-то фича константна (напр. distance_m=0 у всех) — выкидываем её.
    X, kept = _drop_constant_columns(X_full)
    feature_names_full = ("intercept", "rooms", "log_area", "distance_km")
    feature_names = tuple(feature_names_full[i] for i in kept)
    if X.shape[1] < 2:
        logger.info("hedonic: после dedupe осталась только константа — пропускаю")
        return None
    if np.linalg.matrix_rank(X) < X.shape[1]:
        logger.info("hedonic: rank-deficient design matrix — пропускаю")
        return None

    coef_reduced, _, rank, _ = np.linalg.lstsq(X, y, rcond=None)
    if rank < X.shape[1]:
        logger.info("hedonic: rank %d < k %d — пропускаю", rank, X.shape[1])
        return None

    # R² и residual std
    y_pred = X @ coef_reduced
    ss_res = float(np.sum((y - y_pred) ** 2))
    ss_tot = float(np.sum((y - y.mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot if ss_tot > 0 else 0.0
    dof = n - X.shape[1]
    residual_std = math.sqrt(ss_res / dof) if dof > 0 else 0.0

    # Восстанавливаем полный вектор коэффициентов (нули для отброшенных фич),
    # чтобы клиенту было удобно мапить на feature_names_full.
    full_coef = [0.0] * len(feature_names_full)
    for src, idx in enumerate(kept):
        full_coef[idx] = float(coef_reduced[src])

    x_target_full = np.array(
        [1.0, float(target.rooms), math.log(target.area_sqm), target.distance_km],
        dtype=float,
    )
    log_pred = float(x_target_full[kept] @ coef_reduced)
    pred = math.exp(log_pred)

    half_log = CI_Z_95 * residual_std
    ci_lo = math.exp(log_pred - half_log)
    ci_hi = math.exp(log_pred + half_log)

    return HedonicResult(
        predicted_price_per_sqm=pred,
        ci_lo_95=ci_lo,
        ci_hi_95=ci_hi,
        n_used=n,
        r_squared=r2,
        residual_std=residual_std,
        feature_names=feature_names,
        coefficients=tuple(full_coef[i] for i in kept),
    )
